### PR TITLE
Improvement of the FFD capabilities

### PR DESCRIPTION
**Pull request by fpalacios**
This implementation allows the definition of cartesian, cylindrical and spherical FFD boxes. The gradient evaluation itself has been also revisited to reduce the required time by skipping the surface deformation process when the variable is has been fixed by the user.

### CODE REVIEW

**Comment by economon** [comment]
This is going to be pretty cool, Francisco. I think we'll need to add some info to the tutorials about how to use this one... maybe in here: https://github.com/su2code/SU2/wiki/Constrained-Optimal-Shape-Design-of-a-Fixed-Wing.

**Comment by talbring** [comment]
Thanks Francisco for this contribution! Although the FFD Framework is already working quite well, there is still (like always) room for improvement. In fact, I am currently working on using BSplines instead of Bezier curves. But this shouldn't affect this one I hope. Do you have by any chance a simple test case I could use to check this ?

File: Common/src/grid_movement_structure.cpp [hunk 1]
```diff
@@ -3895,79 +4234,104 @@ void CSurfaceMovement::SetFFDCPChange_2D(CGeometry *geometry, CConfig *config, C
...
-	string design_FFDBox;
-
+bool CSurfaceMovement::SetFFDCPChange(CGeometry *geometry, CConfig *config, CFreeFormDefBox *FFDBox, CFreeFormDefBox **ResetFFDBox,
+                                      unsigned short iDV, bool ResetDef) {
+  
+  su2double movement[3] = {0.0,0.0,0.0}, Ampl;
+  unsigned short index[3], i, j, k, iPlane, iFFDBox;
+  bool CheckIndex;
+  string design_FFDBox;
+  su2double Scale = config->GetFFD_Scale();
```
Comment (economon) [inline Common/src/grid_movement_structure.cpp]: What are you using the new scale factor for with the FFD?

File: Common/src/grid_movement_structure.cpp [hunk 2]
```diff
@@ -4578,6 +5288,58 @@ void CSurfaceMovement::SetHicksHenne(CGeometry *boundary, CConfig *config, unsig
   
 }
 
+void CSurfaceMovement::SetSurface_Bump(CGeometry *boundary, CConfig *config, unsigned short iDV, bool ResetDef) {
```
Comment (economon) [inline Common/src/grid_movement_structure.cpp]: It looks like this just separates the existing capability to use a HH bump on an open surface from the HH bump routine for airfoils, correct? Or is there more to this change?

File: Common/src/matrix_structure.cpp [hunk 3]
```diff
@@ -1040,13 +1142,45 @@ void CSysMatrix::InverseDiagonalBlock(unsigned long block_i, su2double *invBlock
       invBlock[jVar*nVar+iVar] = aux_vector[jVar];
   }
   
+  //  su2double Det, **Matrix, **CoFactor;
```
Comment (economon) [inline Common/src/matrix_structure.cpp]: Do we need this commented section any longer?

File: Common/src/matrix_structure.cpp [hunk 4]
```diff
@@ -1059,6 +1193,38 @@ void CSysMatrix::InverseDiagonalBlock_ILUMatrix(unsigned long block_i, su2double
...
```
Comment (economon) [inline Common/src/matrix_structure.cpp]: Also, is this commented section needed?

File: SU2_CFD/src/driver_structure.cpp [hunk 5]
```diff
@@ -508,2053 +547,2716 @@ void CDriver::Postprocessing() {
...
-  if (wave) delete integration_container[WAVE_SOL];
-  if (heat) delete integration_container[HEAT_SOL];
-  if (fem) delete integration_container[FEA_SOL];
-  
-  /*--- DeAllocate solution for adjoint problem ---*/
-  if (adj_euler || adj_ns || disc_adj) delete integration_container[ADJFLOW_SOL];
-  if (adj_turb) delete integration_container[ADJTURB_SOL];
-  
-  
+void CDriver::Integration_Postprocessing(CIntegration **integration_container,
```
Comment (economon) [inline SU2_CFD/src/driver_structure.cpp]: Thanks for all of the clean up in these routines.

File: SU2_PY/SU2/opt/scipy_tools.py [hunk 6]
```diff
@@ -142,7 +142,241 @@ def scipy_slsqp(project,x0=None,xb=None,its=100,accu=1e-10,grads=True):
     # Done
     return outputs
     
-    
+# -------------------------------------------------------------------
```
Comment (economon) [inline SU2_PY/SU2/opt/scipy_tools.py]: Nice! Thanks for putting in the other optimizers again.

File: Common/include/config_structure.hpp [hunk 7]
```diff
@@ -1353,24 +1372,12 @@ class CConfig {
 	 * \return Value of the limits for the sections.
 	 */
 	su2double GetSection_Location(unsigned short val_var);
-
```
Comment (fpalacios) [inline Common/include/config_structure.hpp]: Thanks a lot for your comments and, for taking the time to review the implementation. Code style (including indentation) is very important to keep the source code under control. I have fixed the indentation of this file but, maybe, before releasing SU2 the entire indentation of the software should be updated.

File: Common/include/config_structure.hpp [hunk 8]
```diff
@@ -2872,6 +2885,48 @@ class CConfig {
 	 * \return Number of implicit smoothing iterations.
 	 */
 	unsigned short GetMG_CorrecSmooth(unsigned short val_mesh);
+  
+  /*!
+   * \brief Get the number of pre-smoothings in a multigrid strategy.
```
Comment (fpalacios) [inline Common/include/config_structure.hpp]: Oops, you are right. Fixing the comments about FFD in the config file right now.

**Comment by economon** [comment]
@fpalacios: just let me know when you're happy with this one and we'll merge it in.

File: Common/include/matrix_structure.hpp [hunk 9]
```diff
@@ -239,6 +239,29 @@ class CSysMatrix {
 	 */
 	void DeleteValsRowi(unsigned long i);
   
+  /*!
+   * \brief Recursive definition of determinate using expansion by minors. Written by Paul Bourke
```
Comment (fpalacios) [inline Common/include/matrix_structure.hpp]: Looking at Paul Bourke web page, he says "Any source code found here may be freely used provided credits are given to the author." so we are OK. Anyway, these are standard algorithms that we can find in any algebra book.

File: Common/src/grid_movement_structure.cpp [hunk 10]
```diff
@@ -3530,84 +3769,139 @@ void CSurfaceMovement::CheckFFDIntersections(CGeometry *geometry, CConfig *confi
...
+      if ((config->GetFFD_Fix_KDir(index) <= nDegree) && (config->GetFFD_Fix_KDir(index) >= 0))
+        FFDBox->Set_Fix_KPlane(config->GetFFD_Fix_KDir(index));
+    
+  }
   
-  if (KPlane_Intersect_A) { FFDBox->Set_Fix_KPlane(0); FFDBox->Set_Fix_KPlane(1); }
-  if (KPlane_Intersect_B) { FFDBox->Set_Fix_KPlane(nDegree); FFDBox->Set_Fix_KPlane(nDegree-1); }
+  if (config->GetFFD_Continuity() == DERIVATIVE_NONE) {
+    if (rank == MASTER_NODE)
+      cout << "SU2 is fixing the planes to maintain a continuous surface." << endl;
```
Comment (fpalacios) [inline Common/src/grid_movement_structure.cpp]: You are right, I have improved the description in the config template.

File: Common/src/grid_movement_structure.cpp (hunk 1 above)
Comment (fpalacios) [inline Common/src/grid_movement_structure.cpp]: Long story short. This scale factor is a method to relate the value of the design variable with a meaningful movement of the control points (in a design process). It is more or less equivalent to the scale factor that we apply in OPT_OBJECTIVE or OPT_CONSTRAINT.

File: Common/src/grid_movement_structure.cpp (hunk 2 above)
Comment (fpalacios) [inline Common/src/grid_movement_structure.cpp]: Exactly, there are some design problem where we just want to add a bump in a surface (not an airfoil with chord=1) . Currently we are using HH but in the future we should define this in a more general sense, including 3D (for now, it is a place holder for future improvements).

File: Common/src/matrix_structure.cpp (hunk 3 above)
Comment (fpalacios) [inline Common/src/matrix_structure.cpp]: That is an alternative method to compute the inverse of a block. I have had a lot os problems with the ILU0 preconditioned for grid deformation... in ~75% of my problems it doesn't work... and Discrete adjoint doesn't work either. I thought the problem was in some kind of singular matrix, that is the reason I implement some alternatives to existing subroutines. The commented one is more elegant, and we will probably use that in the future. I'll create a new branch to work on the problem.

File: Common/src/matrix_structure.cpp (hunk 4 above)
Comment (fpalacios) [inline Common/src/matrix_structure.cpp]: Same stuff... the code is going to be useful in the near term future... I'll create a branch to study what is happening with the ILU 0 preconditioner and why it is not working properly.

File: SU2_PY/SU2/opt/scipy_tools.py (hunk 6 above)
Comment (fpalacios) [inline SU2_PY/SU2/opt/scipy_tools.py]: A have had a lot of problems with the constrained optimization, but the unconstrained methods are really great! You have probably noticed that the design variable bounds are inside the SU2 grid deformation subroutine... so you can run unconstrained methods with bounds in the design variables.

**Comment by fpalacios** [comment]
Hi Tom, I have pushed a final version of the source code. Thanks!

Francisco

**Pull request merged**